refactor(compiler-cli): Don't extract type implementations

JeanMeche opened the pull request:
This is useful for cases like class `FormControl` is defined as `interface` but the actual implementation uses a private type  `export const FormControl: ɵFormControlCtor`.

fixes #56144

Demo: https://ng-dev-previews-fw--pr-angular-angular-56489-adev-prev-7lmdybs8.web.app/api/forms/FormControl#usage-notes

JoostK commented on packages/compiler-cli/src/ngtsc/reflection/src/typescript.ts:
```diff
@@ -398,9 +399,14 @@ export class TypeScriptReflectionHost implements ReflectionHost {
+    // Value declarations are given precedence over type declarations if there not referencing a private type
+    if (
+      symbol.valueDeclaration !== undefined &&
+      symbolType?.symbol?.name.startsWith('ɵ') !== true
```
This is in the main compiler logic, not just the docs extraction, where I feel uncomfortable having logic like this.

JoostK replied:
With that option, do we need to make this behavior conditional on the ɵ-prefix? Could we always use the first declaration the flag is set?

JeanMeche replied:
Do you have any insight why the compile behaves that way in the first place ? (prefering the valueDeclaration when there is one.) ?

JoostK replied:
That's typically the most valuable one, i.e. it is the one with decorators if there are any, so any compiler metadata will also have been recorded using the value declaration.

JeanMeche replied:
Perfect, just found out that my change without the `startsWith('ɵ')` guard exactly broke this kind of implementation.

JoostK reviewed (commented):
This needs a test (e.g. in `https://github.com/angular/angular/blob/0d78a92431178be73342aeddda4b62d9e49138ba/packages/compiler-cli/test/ngtsc/doc_extraction/doc_extraction_filtering_spec.ts#L20`) and ideally also a comment in `getDeclarationOfSymbol` with background on why this flag exists.

JoostK commented on packages/compiler-cli/src/ngtsc/reflection/src/typescript.ts:
```diff
@@ -135,7 +135,10 @@ export class TypeScriptReflectionHost implements ReflectionHost {
-  getExportsOfModule(node: ts.Node): Map<string, Declaration> | null {
+  getExportsOfModule(
+    node: ts.Node,
+    skipPrivateValueDeclarationTypes: boolean = false,
```
The compiler typically avoids using defaulted args to make every caller consider the value instead of assuming the default. This case is a bit debatable given that the flag isn't typically desired, so from that perspective I can see how keeping the default makes sense, but it does make me wonder if this is more appropriate as a class field passed via the constructor.

JoostK commented on packages/compiler-cli/src/ngtsc/reflection/src/typescript.ts:
```diff
@@ -135,7 +141,10 @@ export class TypeScriptReflectionHost implements ReflectionHost {
-  getExportsOfModule(node: ts.Node): Map<string, Declaration> | null {
+  getExportsOfModule(
+    node: ts.Node,
+    skipPrivateValueDeclarationTypes: boolean = false,
```
Can be reverted now

JoostK commented on packages/compiler-cli/src/ngtsc/reflection/src/typescript.ts:
```diff
@@ -30,10 +30,16 @@ import {isNamedClassDeclaration} from './util';
  * reflector.ts implements static reflection of declarations using the TypeScript `ts.TypeChecker`.
  */
 
+/**
```
This ought to be the docblock for the parameter itself, or the constructor alternatively, not the class

JoostK commented on packages/compiler-cli/src/ngtsc/reflection/src/typescript.ts:
```diff
@@ -152,6 +161,7 @@ export class TypeScriptReflectionHost implements ReflectionHost {
     this.checker.getExportsOfModule(symbol).forEach((exportSymbol) => {
       // Map each exported Symbol to a Declaration and add it to the map.
       const decl = this.getDeclarationOfSymbol(exportSymbol, null);
+
```
Please also revert unless it's done by the formatter

JoostK reviewed (approved):
Thanks for all the updates!

AndrewKushnir commented:
This PR was merged into the repository by commit 37b88a5a98a2a40e634eb28caf52367cccae3575.

The changes were merged into the following branches: main, 18.2.x

AndrewKushnir commented:
The change will be reverted via PR https://github.com/angular/angular/pull/57594 due to internal breakages.

JeanMeche committed:
refactor(compiler-cli): extract function overload separatly

in order for the docs to process function entry, this commit refactor function extraction by keeping the implementation as a the default entry and adds all the overloads into a separate array of entries.

fixes #56144

AndrewKushnir reviewed (commented):
@JeanMeche just added a couple comments. I can run a TGP once the code is updated. Thank you.

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/OCL:710101443:BASE:710100984:1735327403014:5aefc74e).

AndrewKushnir commented:
FYI, the global presubmit is successful, adding this PR back to the merge queue.

AndrewKushnir commented:
**Caretaker note**: the global presubmit for this PR was successful, but since the PR was reverted earlier (the original issue was resolved since that time), it might be best to merge and sync this change separately.

thePunderWoman commented:
This PR was merged into the repository by commit e0401ec1f0f856a7caf8aa93762bf9c369f3f871.

The changes were merged into the following branches: main, 19.0.x
